### PR TITLE
refactor: casl factory - history

### PR DESCRIPTION
**Pull request by HayenNico**
## Description
Subsection of PR #2748 for history. PR depends on #2759 to be merged first.

This unifies the historyEndpointAccess and historyInstanceAccess functions in CaslAbilityFactory, and adjusts the history controller to accommodate the change. The history-specific auth code is extracted into a separate module.

## Changes:

* Replace CaslAbilityFactory.historyInstanceAccess and CaslAbilityFactory.historyEndpointAccess with one function CaslAbilityFactory.historyAccess
* Code for CaslAbilityFactory.hsitoryAccess is factored out into new module HistoryAbility
* Rename endpoint-level action from HistoryReadEndpoint to HistoryRead
* Adjust endpoint and instance auth logic in attachment controller

## Tests included

- [x] Included for each change/fix?
- [x] Passing? 

## Documentation
- [x] swagger documentation updated (required for API changes)
- [x] official documentation updated

## Summary by Sourcery

Unify history authorization into a single CASL history ability and update history endpoints to use subject-based HistoryRead permissions.

Enhancements:
- Extract history authorization logic into a dedicated HistoryAbility service and inject it into CaslAbilityFactory.
- Replace separate history endpoint and instance access methods with a unified historyAccess method in CaslAbilityFactory.
- Refactor history controller to use subject-based HistoryRead checks instead of per-subsystem action mappings and GenericHistory strings.
- Simplify CASL action enum by consolidating history endpoint and instance actions under a single HistoryRead action while preserving per-collection differentiation via subjects.

Build:
- Register the new HistoryAbility provider in the CASL module for dependency injection.

### CODE REVIEW

File: src/casl/abilities/history.ability.ts [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+     */
+    if (!user) {
+      return build({
+        detectSubjectType: (item) =>
+          item.constructor as ExtractSubjectType<Subjects>,
+      });
+    }
+
+    if (
+      user.currentGroups.some((g) =>
```
[the remark this answers is not in the document]
Comment (HayenNico) [reply src/casl/abilities/history.ability.ts:54]: The current code matches behavior in all other casl ability builders. The extra isArray check is redundant with the interface definition of JWTUser

**Comment by HayenNico** [comment]
Re: Admin level access

There appears to be a mismatch between the GET /history and GET /history/count endpoints - the former allows admins to omit the subsystem filter, the latter does not. Should this be fixed here (will make this breaking instead of refactor)?

File: docs/developer-guide/authorization/history.md
Comment (nitrosx) [inline docs/developer-guide/authorization/history.md:1]: Are we missing HISTORY_ACCESS_ORIGDATABLOCKS_GROUPS?

**Review by nitrosx — COMMENTED** [review]
Do we have history on origDatablocks? 
If they do  not have history, I will accept the PR, with the understanding that we need to work on it in the near future

**Comment by nitrosx** [comment]
# Code Review Report: refactor-casl-factory-history Branch

Generated by Mistral Vibe.
Co-Authored-By: Mistral Vibe <vibe@mistral.ai>

---

## Overview

1. **Verify logic correctness**: The refactoring maintains the same authorization logic but implements it more cleanly using CASL's subject typing system. Logic is correct and equivalent to the original implementation.

2. **Check if all edge cases are handled**: Edge cases are properly handled. The code checks for undefined `accessGroups`, validates subsystem parameters, and handles both authenticated and unauthenticated users appropriately.

3. **Summarize what the code touched by the changes does**: The changes refactor history authorization from string-based action types to class-based subject types in CASL. The system controls read access to history records for various subsystems (Attachment, Datablock, Dataset, Instrument, Policy, Proposal, PublishedData, Sample) based on user group membership.

4. **Assess whether the changes make sense**: The changes are a significant improvement. They simplify the action enum by consolidating 9 history-specific actions into one (`HistoryRead`), eliminate code duplication, improve type safety by using actual class types instead of strings, and separate concerns into a dedicated `HistoryAbility` class.

5. **Identify any unreachable code**: No unreachable code identified. The admin check in `HistoryAbility.buildAbility()` includes both `Action.AccessAny` and specific `HistoryRead` permissions, but this is intentional for clarity and consistency.

**Commits responsible**: 30b94eec (initial refactor), 116d9806 (admin permissions), 31b18018 (undefined accessGroups fix), cfeec844 (typing fix), e67a1e96 (lint), 3486fd6e (docs), 7f830b6b (test fixes)

---

## Code Changes

### Files Modified
- `src/casl/action.enum.ts` - Removed 9 history-specific actions, added single `HistoryRead` action
- `src/casl/casl-ability.factory.ts` - Removed 2 history methods (~220 lines), added delegation to `HistoryAbility`
- `src/casl/casl-ability.factory.spec.ts` - Added `HistoryAbility` to test initialization
- `src/casl/casl.module.ts` - Added `HistoryAbility` provider
- `src/history/history.controller.ts` - Updated to use new action/subject model
- `src/casl/abilities/history.ability.ts` - **New file**: 171 lines of centralized history authorization logic
- `docs/developer-guide/authorization/history.md` - **New file**: 80 lines of documentation

### Assessment
**Are the changes necessary?** Yes. The original implementation had several issues:
- String-based subject types ("GenericHistory") instead of actual class types
- 9 separate action types for each subsystem, creating maintenance burden
- Duplicated logic between endpoint and instance access methods
- Inconsistent permission checking patterns

**Improvement Needed**:
- The `GenericHistory` class import in `history.ability.ts` could be typed more specifically
- Consider adding unit tests specifically for `HistoryAbility` class
- The admin permission check grants both `Action.AccessAny` and all specific `HistoryRead` permissions - could be simplified to just `Action.AccessAny` since it's checked in the controller

**Verdict**: The refactoring is well-executed and addresses technical debt. The changes improve code maintainability, type safety, and follow the established pattern used by `DatasetAbility`.

---

## Security Review

1. **Are there any potential injection vulnerabilities?**
   No. The code properly parses JSON input using `JSON.parse()` with error handling. Query parameters are validated before use.

2. **Does this code expose any sensitive user data?**
   No. The history endpoints only return history records that the user is authorized to access based on their group membership.

3. **Are there instances of insecure API usage?**
   No. All endpoints use `@UseGuards(AuthenticatedPoliciesGuard)` and `@CheckPolicies` decorators to enforce authorization before any data access.

4. **Could this code lead to an authentication bypass?**
   No. The refactoring maintains the same security model:
   - Unauthenticated users: no access
   - Authenticated users: no access by default
   - Users in specific groups: access only to authorized subsystems
   - Admin users: full access
   The new implementation uses `ability.can(Action.AccessAny, GenericHistory)` for admin check, which is equivalent to the previous admin checks.

**Commits responsible**: All changes maintain or improve security posture.

---

## Test Coverage

### Current State
- `casl-ability.factory.spec.ts` updated to include `HistoryAbility` in initialization tests
- No dedicated unit tests for `HistoryAbility` class itself
- No integration tests for the new permission model in `HistoryController`

### Assessment
Test coverage is **inadequate** for the new functionality. While the factory initialization is tested, there are no tests that:
1. Verify `HistoryAbility.buildAbility()` returns correct permissions for different user types
2. Test the permission matrix described in the documentation
3. Validate edge cases (undefined accessGroups, missing subsystem, invalid subsystem)

### Suggestions for Improvement
1. Create `history.ability.spec.ts` with tests for:
   - Unauthenticated user returns empty ability
   - Users in each specific group get appropriate permissions
   - Admin users get all permissions
   - Multiple group memberships are handled correctly

2. Add integration tests in `history.controller.spec.ts`:
   - Test each subsystem endpoint with users having appropriate permissions
   - Test that users without permissions receive 403
   - Test that invalid subsystems return 400

3. Test the `Action.AccessAny` check for admin users in the controller

**Relevant commits**: 7f830b6b (test initialization fix), but comprehensive tests still missing

---

## Security Examples

No security vulnerabilities were identified in the review. The refactoring maintains the same security model as the original implementation.

---

## Testing for Security Use Cases

### Authentication Bypass Testing
**Test Case**: Verify unauthenticated users cannot access any history endpoint
```typescript
// Test in history.controller.spec.ts
it('should deny access to unauthenticated users', async () => {
  const response = await request(app.getHttpServer())
    .get('/history?filter={"subsystem":"Dataset"}')
    .expect(401); // Or 403 depending on guard configuration
});
```

**Code affected**: `history.ability.ts:46-50` (unauthenticated user check)

### Authorization Bypass Testing
**Test Case**: Verify users without specific group membership cannot access restricted subsystems
```typescript
// Test in history.ability.spec.ts
it('should deny Dataset history access to users not in historyDataset group', () => {
  const user = { currentGroups: ['some-other-group'] };
  const ability = historyAbility.buildAbility(user);
  expect(ability.can(Action.HistoryRead, DatasetClass)).toBe(false);
});
```

**Code affected**: `history.ability.ts:77-87` (dataset permission check)

### Permission Escalation Testing
**Test Case**: Verify admin users have access to all subsystems
```typescript
it('should grant all history access to admin users', () => {
  const user = { currentGroups: ['admin-group'] };
  const ability = historyAbility.buildAbility(user);
  expect(ability.can(Action.AccessAny, GenericHistory)).toBe(true);
  expect(ability.can(Action.HistoryRead, DatasetClass)).toBe(true);
  // ... test all subsystems
});
```

**Code affected**: `history.ability.ts:149-164` (admin permission grants)

---

## Summary

The `refactor-casl-factory-history` branch introduces a well-designed refactoring of the history authorization system. The changes consolidates 9 action types into one, improves type safety by using class-based subjects instead of strings, and separates concerns into a dedicated `HistoryAbility` class following the established pattern.

**Key Findings:**
- **Logic Correctness**: Maintained - new implementation is functionally equivalent
- **Security**: No regressions - same authorization model preserved
- **Code Quality**: Significantly improved - better type safety, less duplication
- **Test Coverage**: Inadequate - needs dedicated tests for `HistoryAbility` and integration tests for the new model
- **Documentation**: Excellent - comprehensive new documentation file

**Recommendations:**
1. Add unit tests for `HistoryAbility` class
2. Add integration tests for the refactored `HistoryController`
3. Consider simplifying admin permissions to use only `Action.AccessAny`
4. Verify the new pattern works correctly in production-like environments

**Attribution**: This review was generated by Mistral Vibe based on analysis of commits 30b94eec, 116d9806, 31b18018, cfeec844, e67a1e96, 3486fd6e, and 7f830b6b in the refactor-casl-factory-history branch.

File: docs/developer-guide/authorization/history.md
Comment (HayenNico) [reply docs/developer-guide/authorization/history.md:1]: Origdatablocks are not currently available for history tracking after a bit of research - it wouldn't be too difficult to add them (need to add the `addHistoryPluginOnce` call at the module level), but that should be separate from this PR.

**Review by nitrosx — APPROVED** [review]
